app/models/category_list.rb

```diff
@@ -138,14 +139,32 @@ def find_categories
 
     if SiteSetting.lazy_load_categories
       page = [1, @options[:page].to_i].max
-      query = query.limit(CATEGORIES_PER_PAGE).offset((page - 1) * CATEGORIES_PER_PAGE)
+      query =
+        query
+          .where(parent_category_id: nil)
+          .limit(CATEGORIES_PER_PAGE)
+          .offset((page - 1) * CATEGORIES_PER_PAGE)
     end
 
     query =
       DiscoursePluginRegistry.apply_modifier(:category_list_find_categories_query, query, self)
 
     @categories = query.to_a
 
+    if SiteSetting.lazy_load_categories
+      categories_with_rownum =
+        Category
+          .secured(@guardian)
+          .select(:id, "ROW_NUMBER() OVER (PARTITION BY parent_category_id) rownum")
+          .where(parent_category_id: @categories.map { |c| c.id })
+
+      @categories +=
+        Category.where(
+          "id IN (WITH cte AS (#{categories_with_rownum.to_sql}) SELECT id FROM cte WHERE rownum <= ?)",
```

Comment:
Maybe something for the future but I wonder if we need to think about ordering? For example, we might want to display the top 5 most subcategories as the default

Reply:
I don't think we should use a CTE here, since it forces the set to be materialized. That could be expensive if a category has a huge number of subcategories.